gtsfm/densify/mvs_patchmatchnet.py

```diff
@@ -0,0 +1,251 @@
+
+        # For each reference view and the corresponding source views
+        for pair in packed_pairs:
+            ref_view = pair["ref_id"]
+            src_views = pair["src_ids"]
+
+            # Load the camera parameters
+            ref_intrinsics, ref_extrinsics = dataset.get_camera_params(ref_view)
+
+            # Load the reference image
+            ref_img = dataset.get_image(ref_view)
+            # Load the estimated depth of the reference view
+            ref_depth_est = depth_list[ref_view][0]
+            # Load the photometric mask of the reference view
+            confidence = confidence_list[ref_view]
+            # Filter the pixels that satisfy photometric consistancy among reference view and source views,
+            #   by checking whether the confidence is larger than the pre-defined photometric threshold
+            photo_mask = confidence > photo_thresh
+
+            all_srcview_depth_ests = []
+
+            # Compute the geometric mask, the value of geo_mask_sum means the number of source views where
+            #   the reference depth is valid according to the geometric thresholds
+            geo_mask_sum = 0
+            for src_view in src_views:
+                # camera parameters of the source view
+                src_intrinsics, src_extrinsics = dataset.get_camera_params(src_view)
+
+                # the estimated depth of the source view
+                src_depth_est = depth_list[src_view][0]
+
+                # Check geometric consistency
+                geo_mask, depth_reprojected, _, _ = check_geometric_consistency(
+                    ref_depth_est,
+                    ref_intrinsics,
+                    ref_extrinsics,
+                    src_depth_est,
+                    src_intrinsics,
+                    src_extrinsics,
+                    geo_pixel_thresh,
+                    geo_depth_thresh,
+                )
+                geo_mask_sum += geo_mask.astype(np.int32)
+                all_srcview_depth_ests.append(depth_reprojected)
+
+            depth_est_averaged = (sum(all_srcview_depth_ests) + ref_depth_est) / (geo_mask_sum + 1)
+            # Valid points requires at least 3 source views validated under geometric threshoulds
+            geo_mask = geo_mask_sum >= 3
+
+            # Combine geometric mask and photometric mask
+            final_mask = np.logical_and(photo_mask, geo_mask)
+
+            # Initialize coordinate grids
+            height, width = depth_est_averaged.shape[:2]
+            x, y = np.meshgrid(np.arange(0, width), np.arange(0, height))
+
+            # Get valid points filtered by photometric and geometric thresholds
+            valid_points = final_mask
+            x, y, depth = x[valid_points], y[valid_points], depth_est_averaged[valid_points]
+
+            # Get the point coordinates in world frame
+            xyz_ref = np.matmul(np.linalg.inv(ref_intrinsics), np.vstack((x, y, np.ones_like(x))) * depth)
```

Comment:
i think 
```python
def cart_to_homogenous(non_homogenous_coordinates: np.ndarray,) -> Optional[np.ndarray]:
    """Convert cartesian coordinates to homogenous system (by appending a column of ones). 
   Args: 
         non_homogenous_coordinates: 3d non-homogenous coordinates, of shape Nx3.
    
    Returns: 
          4d homogenous coordinates, of shape Nx4. 

    Raises: TypeError: if input is not 2 dimensional. 
     """
     n = non_homogenous_coordinates.shape[0]
     return np.hstack([non_homogenous_coordinates, np.ones((n, 1))])  

# get the point coordinates inside the reference view's camera frame
xyz_ref = np.linalg.inv(ref_intrinsics) @ np.vstack([u, v, np.ones_like(u])) * depth)
# get the point coordinates inside the world frame
xyz_world = wTref @ cart_to_homogeneous(xyz_ref)
```
is more readabile

Reply:
let's call it u, v, not x, y?

Reply:
shall I put `cart_to_homogenous` into `utils/geometry_comparison`?

Reply:
Let's just put it as a helper in this file for now. I don't think it's so related to geometric comparisons

Reply:
much better now, thanks